### ISSUE TITLE
Rcm(Core): Implement PSR7 Middle-Ware (zend-expressive?)

### ISSUE THREAD

**Issue by jerv13** [reliv/rwriter#28]
- As time permits (not a priority)
    - Need more investigation 
    - Limit BC breaks

**Comment by rodmcnew** [comment]
If this BC-breaks any interfaces, I think we should consider having an "PSR7 to old interface" adaptor that allows the old interfaces to keep functioning. This could sit on top of the PSR7 layer and be simply deleted later. Maybe this would be a separate module.

**Comment by wshafer** [comment]
ZF 3 I think is the better way forward.  They will have a PSR-7 middleware layer, which will let us move some of our more middleware components out to to the middleware stack and allow us to shift our services to the container-interop.   In addition they already have a PSR-7 layer to Zend response/request attached.

As I have already had some experience with moving the core to expressive let me share what I've discovered.  

While some things work better, we have a lot of dependencies on ZF (mainly the MVC stack) that I think we all forget are there.  Because of that one of the big issues I've had with Expressive, was I was solving a lot of issues that are already solved in ZF.  

Expressive has no module system, yet.  While ZF was built with modules from the ground up.  This allows us to have things like Plugins with little to no effort.   Currently this is not completely possible yet in expressive.

Another issue is in the differences between the view stacks.  For a move to expressive to work, we would still have a hard dependency on Zend-view.  The other two view layers appear to render from the top-down, when what we actually need is for them to render from the bottom up.  That allows the plugins to be more independent of the core, and at the same time allows the core to be very aggressive with it caching.  Yes we could solve this by making each plugin actually pass themselves completely rendered back to the core, how do we manage things like CSS and JS?  Right now at least plugins can use the whole view stack in zf 2 and for the most part Core can handle this.

I do agree, some of our services are better fits for the middleware stack.  These should definitely be addressed in ZF3 where we'll have a middleware layer we can attach to.

For the record, I am not opposed at all to moving out to something a bit more flexible like expressive, or slim, or [enter micro-framework here].  But I do think ZF 3 will be a good transition to move us in this direction.  Perhaps by then we'll have better abstractions for the view layers and a standardized module and packaging system.  I think it's only a matter of time.

**Comment by wshafer** [comment]
Personally, I'd love to see a PSR-8 that extends PSR-7 for the html layer, so we can build-up html responses in a middleware fashion.  Then a move to expressive would be trivial.  ;)

**Comment by rodmcnew** [comment]
Some notes on service containers:

Container-interop by its self does basically nothing. It is ONLY an interface. Services still must be defined using config language for a certain container whether that be ZF2, Symfony, or https://github.com/reliv/zf-config-factories. It may be nice to use ContainerInterface instead of ServiceManager in your type hints but that is ALL this accomplishes.

I do have the https://github.com/rodmcnew/symfonize-zf-container-bridge lib that I built that allows ZF2 and Symfony2's containers to get services from each other but most people agree that this is scary and is a bad idea to use.

So: A single service container needs to be picked. I personally think the ZF2 service container by itself results in extra work and code (factory classes). BUT we have already used it everywhere, AND it is closely related to much of ZF. Those reasons plus the fact that zf-config-factories resolve all my issues with it make me feel too lazy to push to get rid of it.

**Comment by jerv13** [comment]
Yes, it is low priority for sure and the details need to be worked out.  

HTML is a layer below middleware as requests and responses are not exclusively HTML.

**Comment by rodmcnew** [comment]
I highly question the current system of passing view models everywhere rather than PSR7 responses. I've wondered for a while if the plugin interface would be better off if the plugins could return fully rendered PSR7 responses.

Though if we were to ever do that, I think we would want to allow the old plugin interface to keep operating in some way. Perhaps core could allow more than one "plugin system" to register it's self? I don't know.

**Comment by rodmcnew** [comment]
I agree with @wshafer that if expressive has no module system, that is a huge red flag for using it.

**Comment by rodmcnew** [comment]
I think its worth considering the idea that both plugins and the CMS itself could be exposed through this interface https://github.com/zendframework/zend-stratigility/blob/master/src/MiddlewareInterface.php

Maybe this interface doesn't fit, but if it does, it really standardizes things.

**Comment by rodmcnew** [comment]
I think @jerv13 more meant zend-stratigility than zend-expressive so I changed the title. Zend-stratigility contains all the PSR7 middleware code, zend-expressive is just a micro-framework built on top of it.

I agree with @wshafer that using a microframework for RCM is most likely a bad idea due to all the missing features in microframeworks.

**Comment by rodmcnew** [comment]
>Yes we could solve this by making each plugin actually pass themselves completely rendered back to the core, how do we manage things like CSS and JS?

@wshafer Good question.

**Comment by jerv13** [comment]
Yes, the PSR7 layer is just and abstract way of sharing standard Requests and Responses.  Does not solve any Controller, ViewModel, ServiceManager, or Event issues, these are still tightly coupled to ZF.
It only allows alternate code and frameworks to be used in the same code base, but completely decoupled from RCM and ZF.

How to share Controllers, ViewModels, ServiceManagers (or services), or Events is a whole other box of issues that will need to be considered.

**Comment by wshafer** [comment]
I have a couple of points with above comments.

First container-interop, yes is only an interface.  But all containers will work using it.  In addition, the configuration is also dictated to a point.  If you take a look at how expressive does this you'll see what I mean.  The only downside is that by using container-interop, in the manner Zend has implemented it in Expressive, you loose the abstract factories and more complex service creation from the Service Manager.  Only invokables and factories would work.

I completely disagree with the plugins returning PSR7 with fully rendered responses.  In PSR 7 there is no way to pull out sections of the HTML.  So the only way to compose a full HTML response is to parse the returning response.

The other option is to build our own HttpResponse object that extends PSR7, but then we're just reinventing Zend-View with ViewModels.   Plus, we don't seem to document things very well, so how would one even work with our own rendering system.

The best approach is still to use a standard system for this.

@jerv13 - In response to your last comment, doesn't ZF3, Symfony 3 already address this with their middleware layer?

**Comment by jerv13** [comment]
I think ZF3 is a good direction as long as BC breaks are limited.

**Comment by jerv13** [comment]
Also, this issue is really to discuss possible directions, no hard decisions have been made.

**Comment by rodmcnew** [comment]
@wshafer Looks like I was confused. Apparently @jerv13 's goal with this ticket was to allow other PSR middleware apps to be able to run in an RCM app and to encourage decoupling. This issue doesn't seem to involve major changes to RCM its self, or plugin interfaces, or service containers.

From what I can tell, ZF3 would meet this goal of allowing other PSR7 middleware apps to be registered and run inside of an RCM app. And with ZF3, we wouldn't have to build this functionality our selves.

**Comment by wshafer** [comment]
BTW.  I don't mean to be so negative here.  I like the idea of moving to some of these standards.   I also liked the freedom of Expressive.  However there's a lot of road blocks to address before that can be done, and none of that without massive BC breaks.   The other thing I came to the conclusion of, even if you went with Expressive under the hood, in it's current form we would still have to dictate the rendering engine, since these are not standardized.

But there is a lot to like about this RFC.  For one, almost ALL of cores event listeners are a much better fit for the Middleware pattern over the event system.  Those need to fire almost immediately. Why fire up the framework if we're just going to redirect them?

Api's like the ones in Rcm-Admin and Rcm-User are also perfect fits for the middleware layer.

**Comment by wshafer** [comment]
After reading your guys last comments, it looks like we're all on the same page now.  Now all we can do is wait for ZF3....  I hate waiting.  lol

**Comment by rodmcnew** [comment]
+1 on avoiding the ZF event manager in favor of other alternatives such as middleware.

**Comment by jerv13** [comment]
+100 on avoiding the ZF event manager in favor of other alternatives such as middleware.

**Comment by rodmcnew** [comment]
The dictation of the rendering engine was a big problem I ran into with my project to allow plugins that are Symfony bundles (this project has been abandoned). I think it could be good if plugins could return fully rendered responses if they want to. If we allow fully rendered PSR7 responses, then people can use any template system they want inside any plugin.

**Comment by rodmcnew** [comment]
@wshafer 
> I completely disagree with the plugins returning PSR7 with fully rendered responses. In PSR 7 there is no way to pull out sections of the HTML. So the only way to compose a full HTML response is to parse the returning response.

I missed this comment earlier and just saw it. Yes PSR7 responses would be a little different from the view models we have now. I don't see why the plugin engine couldn't say "if I get a 200 OK response from a plugin, just shove it on the page where that plugin goes". Perhaps this goes against the way much of the core is currently written. 

I currently don't care much about plugins returning PSR7 responses. Was just an idea to throw out there. I've found that many frameworks such as Symfony have simplified controllers compared to ZF, they take a Request in their action's arguments and simply return a Response. This is where I got the idea. Since I don't care much about this idea, I'll leave it be. I can always push for it in it's own github issue later.

**Comment by wshafer** [comment]
Sorry not to beat a dead horse, but I'm curious about your thoughts on this.  What if $response->getBody() returns:

```html
<html>
<head>
   <!-- lots of js and css here -->
</head>
<body>
   <!-- Content -->
</body>
<footer>
   <!-- Footer -->
</footer>
</html>
```
How do I get this info back to the layout template (which would have to be some standard like Twig, Zend-View, etc)

**Comment by wshafer** [comment]
@jerv13 & @rodmcnew - Perhaps we don't have to wait for ZF3.  See this doc, it's very interesting  http://tomhanderson.blogspot.com/2015/11/graft-zend-framework-2-on-to-any-legacy.html

**Comment by rodmcnew** [comment]
@wshafer 
> What if $response->getBody() returns:
We could say plugins are not allowed to return <html>, <body>, or <head> tags but that would leave the question of how they could inject their own css (they can inject css through the asset manager combined file but not everyone may want to do this).

Another option could be to look for <head> in the response from plugins and just append anything in there to the actual head. Duplicates could be a complication here.

I see your point that PSR7 responses from plugins could get messy.

**Comment by rodmcnew** [comment]
@wshafer
>Perhaps we don't have to wait for ZF3

That article looks along the lines of what James wanted to investigate. We don't have any immediate needs for this though, so this could probably wait till ZF3 has middleware registry built in.

**Comment by wshafer** [comment]
It works.  Here's my index.php file now

```php
<?php

use Zend\Stratigility\MiddlewarePipe;
use Zend\Diactoros\Server;

/**
 * This makes our life easier when dealing with paths. Everything is relative
 * to the application root now.
 */
chdir(dirname(__DIR__));

// Setup autoloading
require 'init_autoloader.php';

// Decline static file requests back to the PHP built-in webserver
if (php_sapi_name() === 'cli-server') {
    $path = realpath(__DIR__ . parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH));
    if (__FILE__ !== $path && is_file($path)) {
        return false;
    }
    unset($path);
}

$skipEnv = getenv('SKIP_ENV') ? getenv('SKIP_ENV') : false;

if (!$skipEnv) {
    $dotenv = new Dotenv\Dotenv('.');
    $dotenv->load();
}

$app    = new MiddlewarePipe();
$server = Server::createServer($app, $_SERVER, $_GET, $_POST, $_COOKIE, $_FILES);

$bootstrapZF2 = function()
{
    // Run the application!
    Zend\Mvc\Application::init(require 'config/application.config.php')->run();
    exit;
};

$server->listen($bootstrapZF2);
```

**Comment by rodmcnew** [comment]
Nice example. Looks like it will work if you want to inject middleware before the ZF2 MVC, but middleware cannot be injected after since the ZF2 MVC itself is not a middleware.

**Comment by wshafer** [comment]
Not yet.  But it will be soon.  ;)

**Comment by jerv13** [comment]
One word... PipeRat
